### PR TITLE
add test and fix #3254

### LINKED ISSUE #3254: Async computation hanging in lastest F# 4.1 builds
While going through the discussion in #3219 I noticed that a particular computation seems to be hanging when run in F# Interactive 4.1

cc @matthid

#### Repro steps

In F# Interactive 4.1
```fsharp
open System
open System.Threading
open System.Threading.Tasks

let test() =
    let tcs = new TaskCompletionSource<int>()
    let cts = new CancellationTokenSource()
    let _ = cts.Token.Register(fun () -> tcs.SetResult 42)
    async {
        cts.CancelAfter 500
        let! result = tcs.Task |> Async.AwaitTask
        return result
    } |> fun a -> Async.RunSynchronously(a, cancellationToken = cts.Token)
```
function test does execute as expected. 

#### Expected behavior

Should fail with OperationCanceledException

#### Actual behavior

Hangs indefinitely

#### Related information 

This only occurs in F# interactive 4.1 (32bit and 64bit builds) in both windows and mono/linux. It does not happen in console applications running FSharp.Core 4.4.1.0. I can't reproduce the issue in F# Interactive 4.0.

### PR DESCRIPTION
**Pull request by matthid**
See https://github.com/Microsoft/visualfsharp/issues/3254
/cc @eiriktsarpalis

### CODE REVIEW

**Comment by eiriktsarpalis** [comment]
@matthid good catch, this should never have been added. Could also remove the `continueWithExtra` definition?

File: src/fsharp/FSharp.Core.Unittests/FSharp.Core/Microsoft.FSharp.Control/AsyncType.fs [hunk 1]
```diff
@@ -162,7 +162,22 @@ type AsyncType() =
         this.WaitASec t
         Assert.IsTrue (t.IsCompleted)
         Assert.AreEqual(s, t.Result)    
-      
+
+    [<Test>]
+    member this.RunSynchronouslyCancellationWithDelayedResult () =
```
Comment (eiriktsarpalis) [inline src/fsharp/FSharp.Core.Unittests/FSharp.Core/Microsoft.FSharp.Control/AsyncType.fs]: Could you add some kind of timeout to this test? Assuming the fix is not in place that test would run forever. Perhaps configurable via nunit?

**Comment by matthid** [comment]
> @matthid good catch, this should never have been added. Could also remove the continueWithExtra definition?

Yep added.

File: src/fsharp/FSharp.Core.Unittests/FSharp.Core/Microsoft.FSharp.Control/AsyncType.fs (hunk 1 above)
Comment (matthid) [inline src/fsharp/FSharp.Core.Unittests/FSharp.Core/Microsoft.FSharp.Control/AsyncType.fs]: I actually thought about that, but to be honest there are a lot of tests with similar expectations. Not sure a timeout there is worth it, but I can of course add one if you don't agree

Comment (eiriktsarpalis) [inline src/fsharp/FSharp.Core.Unittests/FSharp.Core/Microsoft.FSharp.Control/AsyncType.fs]: as long as it doesn't obscure the intention of the test I guess

Comment (eiriktsarpalis) [inline src/fsharp/FSharp.Core.Unittests/FSharp.Core/Microsoft.FSharp.Control/AsyncType.fs]: https://www.nunit.org/index.php?p=timeout&r=2.5

Comment (matthid) [inline src/fsharp/FSharp.Core.Unittests/FSharp.Core/Microsoft.FSharp.Control/AsyncType.fs]: Thanks, done

**Comment by eiriktsarpalis** [comment]
@matthid this issue seems related https://github.com/Microsoft/visualfsharp/pull/2770#discussion_r123879407 could you also include this here? @saul

**Comment by matthid** [comment]
@eiriktsarpalis Any idea how a test for this would look like or what wrong behavior this would expose?
I can include it here but then there will be no test :)
To be honest I don't even think this code-path is used in any real life code. Only `Async.Sleep` is using it and there the result is correct, isn't it?

**Comment by eiriktsarpalis** [comment]
> To be honest I don't even think this code-path is used in any real life code. Only Async.Sleep is using it and there the result is correct, isn't it?

That's correct. It's splitting hairs really, but I happened to spot this in the same PR you're reverting.

**Comment by matthid** [comment]
@eiriktsarpalis ok perfect, added it :)

**Comment by eiriktsarpalis** [comment]
Cheers, lgtm

**Comment by matthid** [comment]
@eiriktsarpalis It seems like something doesn't like the timeout attribute, ideas?

**Comment by dsyme** [comment]
> What the cancellationToken does at this place is cancel the Task which is RETURNED by ContinueWith which means the continuation is never called. I'd remove this parameter.

Ugh... good catch, and I'm glad we found this now rather than later.

File: src/fsharp/FSharp.Core/control.fs [hunk 2]
```diff
@@ -1536,43 +1536,35 @@ namespace Microsoft.FSharp.Control
...
-            task.ContinueWith(Action<Task<'T>>(continuation), continueWithExtra args.aux.token) |> ignore |> fake
+            task.ContinueWith(Action<Task<'T>>(continuation)) |> ignore |> fake
 
         let continueWithUnit (task : Task, args, useCcontForTaskCancellation) = 
 
             let continuation (completedTask : Task) : unit =
                 args.aux.trampolineHolder.Protect((fun () ->
                     if completedTask.IsCanceled then
                         if useCcontForTaskCancellation
-                        then args.aux.ccont (new OperationCanceledException(args.aux.token))
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/control.fs]: Is this an intrinsic part of this fix?  Or is only https://github.com/Microsoft/visualfsharp/pull/3255/files#diff-34ee82347e9e0b47265fcecaa4850174R1567 needed?  If the latter, please remove this change to keep the changes entirely separate,

File: src/fsharp/FSharp.Core/control.fs [hunk 3]
```diff
@@ -1536,43 +1536,35 @@ namespace Microsoft.FSharp.Control
...
+                        then args.aux.ccont (new OperationCanceledException())
                         else args.aux.econt (ExceptionDispatchInfo.Capture(new TaskCanceledException(completedTask)))
                     elif completedTask.IsFaulted then
                         args.aux.econt (MayLoseStackTrace(completedTask.Exception))
                     else
                         args.cont completedTask.Result)) |> unfake
 
-            task.ContinueWith(Action<Task<'T>>(continuation), continueWithExtra args.aux.token) |> ignore |> fake
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/control.fs]: Is this an intrinsic part of this fix?  Or is only https://github.com/Microsoft/visualfsharp/pull/3255/files#diff-34ee82347e9e0b47265fcecaa4850174R1567 needed?  If the latter, please remove this change to keep the changes entirely separate,

File: src/fsharp/FSharp.Core/control.fs [hunk 4]
```diff
@@ -1536,43 +1536,35 @@ namespace Microsoft.FSharp.Control
...
-#endif
-
         let continueWith (task : Task<'T>, args, useCcontForTaskCancellation) = 
 
             let continuation (completedTask : Task<_>) : unit =
                 args.aux.trampolineHolder.Protect((fun () ->
                     if completedTask.IsCanceled then
                         if useCcontForTaskCancellation
-                        then args.aux.ccont (new OperationCanceledException(args.aux.token))
+                        then args.aux.ccont (new OperationCanceledException())
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/control.fs]: Is this an intrinsic part of this fix?  Or is only https://github.com/Microsoft/visualfsharp/pull/3255/files#diff-34ee82347e9e0b47265fcecaa4850174R1567 needed?  If the latter, please remove this change to keep the changes entirely separate,

**Comment by dsyme** [comment]
My understanding is that only the one line change https://github.com/Microsoft/visualfsharp/pull/3255/files#diff-34ee82347e9e0b47265fcecaa4850174L1575 is required for the fix, the rest is a different issue?

**Comment by matthid** [comment]
The change was requested because it's part of the commit which broke this. I honestly don't care if it is part of this or not as I noted above it's basically changing nothing...

So I don't understand why I had to add it (it's basically cleanup) nor why now I should remove it again :/

So I'm fine either way but don't make me change that back and forth. @dsyme Did you see the above discussion and still think I should remove it?

/cc @eiriktsarpalis

**Comment by dsyme** [comment]
> So I don't understand why I had to add it (it's basically cleanup) nor why now I should remove it again :/

Remove it please to minimise the required change, thanks

**Comment by dsyme** [comment]
@matthid I updated the PR directly, thanks

**Comment by dsyme** [comment]
@KevinRansom this is ready

**Comment by matthid** [comment]
@dsyme thanks. i'd probably rebase it and remove all those commits. but if you are fine with them I can leave it as is

**Comment by matthid** [comment]
Problem is I'm quite busy the mext days so it could take a while

**Comment by dsyme** [comment]
@matthid No problem - we will squash & merge in any case :)

**Pull request merged**